Gems/Atom/RPI/Code/Source/RPI.Reflect/Asset/AssetUtils.cpp

```diff
@@ -33,19 +33,46 @@ namespace AZ
                 }
             }
 
-            Data::AssetId GetAssetIdForProductPath(const char* productPath, TraceLevel reporting, Data::AssetType assetType, bool autoGenerateId)
+            bool TryToCompileAsset(const AZStd::string& assetFilePath, TraceLevel reporting)
             {
+                AzFramework::AssetSystem::AssetStatus status = AzFramework::AssetSystem::AssetStatus_Unknown;
+                AzFramework::AssetSystemRequestBus::BroadcastResult(
+                        status, &AzFramework::AssetSystemRequestBus::Events::CompileAssetSync, assetFilePath);
+
+                if ((status != AzFramework::AssetSystem::AssetStatus_Compiled) && (status != AzFramework::AssetSystem::AssetStatus_Unknown))
+                {
+                    AssetUtilsInternal::ReportIssue(
+                        reporting,
+                        AZStd::string::format(
+                            "Could not compile asset '%s', status = %u.", assetFilePath.c_str(), static_cast<uint32_t>(status))
+                            .c_str());
+
+                    return false;
+                }
+
+                return true;
+            }
```

Comment:
Should this try to lookup the AssetID first using the assetFilePath and if found use `CompileAssetSyncById` in that case?

Reply:
It _could_, but my preferred route would be to phase out the direct use of hard-coded paths entirely and have IDs passed in. (I'll be entering GHIs for these) It also gets a bit tricky because of the way we potentially need _product_ paths for the ID lookup, but _source_ paths for the compile request.

The biggest problem I've seen though with passing in paths is that it prevents us from using the "missing asset" flow and requires the compile to happen to even get an assetId from the lookup. This is especially a problem with the places where azmodel assets are passed in, because they can fail the CompileAssetSync due to their reliance on intermediate assets.

I'll change this now if you feel strongly, but I'd rather see folks work on changing the path usage over to ID usage.

Reply:
What is currently there is fine?

Reply:
What's the benefit of that? I mean using get asset id by file path then use CompileAssetSyncById compare to CompileAssetSync(assetPath)

Reply:
The main benefit is that CompileAssetSync goes through a whole set of heuristics to try to figure out what to compile based on string matching and multiple O(N) searches, where CompileAssetSyncById just does a single O(N) search:
https://github.com/o3de/o3de/blob/development/Code/Tools/AssetProcessor/native/resourcecompiler/rcjoblistmodel.cpp#L393-L563
vs
https://github.com/o3de/o3de/blob/development/Code/Tools/AssetProcessor/native/resourcecompiler/rcjoblistmodel.cpp#L565-L590